File: clients/typescript/src/services/query.ts
import { AxiosInstance } from 'axios';
import { QueryRequest, QueryResponse } from '../types';

export class QueryService {
  constructor(private readonly client: AxiosInstance) {}

  /**
   * Execute a query
   */
  async execute(request: QueryRequest): Promise<QueryResponse> {
    const response = await this.client.post<QueryResponse>('/query', request);
    return response.data;
  }

  /**
   * Validate a query without executing it
   */
  async validate(query: string): Promise<{
    valid: boolean;
    errors?: Array<{
      line: number;
      column: number;
      message: string;
    }>;
    warnings?: string[];
  }> {
    const response = await this.client.post<{
      valid: boolean;
      errors?: Array<{
        line: number;
        column: number;
        message: string;
      }>;
      warnings?: string[];
    }>('/query/validate', { query });
    return response.data;
  }

  /**
   * Get query suggestions based on partial input
   */
  async suggest(partial: string, context?: {
    eventType?: string;
    cursorPosition?: number;
  }): Promise<{
    suggestions: Array<{
      text: string;
      type: string;
      description?: string;
      insertText?: string;
    }>;
  }> {
    const response = await this.client.post<{
      suggestions: Array<{
        text: string;
        type: string;
        description?: string;
        insertText?: string;
      }>;
    }>('/query/suggest', {
      partial,
      ...context
    });
    return response.data;
  }

  /**
   * Format a query
   */
  async format(query: string): Promise<{
    formatted: string;
    changed: boolean;
  }> {
    const response = await this.client.post<{
      formatted: string;
      changed: boolean;
    }>('/query/format', { query });
    return response.data;
  }

  /**
   * Get query history
   */
  async getHistory(options?: {
    limit?: number;
    offset?: number;
    timeRange?: {
      from: string;
      to: string;
    };
  }): Promise<{
    queries: Array<{
      id: string;
      query: string;
      executedAt: string;
      executionTime: string;
      resultCount?: number;
      user?: string;
    }>;
    total: number;
  }> {
    const params = new URLSearchParams();
    
    if (options?.limit !== undefined) {
      params.append('limit', options.limit.toString());
    }
    if (options?.offset !== undefined) {
      params.append('offset', options.offset.toString());
    }
    if (options?.timeRange) {
      params.append('from', options.timeRange.from);
      params.append('to', options.timeRange.to);
    }

    const response = await this.client.get<{
      queries: Array<{
        id: string;
        query: string;
        executedAt: string;
        executionTime: string;
        resultCount?: number;
        user?: string;
      }>;
      total: number;
    }>('/query/history', {
      params: Object.fromEntries(params)
    });
    return response.data;
  }

  /**
   * Export query results
   */
  async export(queryId: string, format: 'csv' | 'json' | 'parquet'): Promise<Blob> {
    const response = await this.client.get(`/query/${encodeURIComponent(queryId)}/export`, {
      params: { format },
      responseType: 'blob'
    });
    return response.data as Blob;
  }
}

File: clients/python/src/newrelic_uds/services/query.py
"""Query service implementation."""

from typing import TYPE_CHECKING, Any, Dict, List, Optional
from urllib.parse import quote

from ..models import QueryRequest, QueryResponse

if TYPE_CHECKING:
    from ..client import AsyncUDSClient, SyncUDSClient


class AsyncQueryService:
    """Async service for query operations."""
    
    def __init__(self, client: "AsyncUDSClient"):
        self.client = client
    
    async def execute(self, request: QueryRequest) -> QueryResponse:
        """Execute a query."""
        return await self.client.post(
            "/query",
            json=request.model_dump(by_alias=True, exclude_none=True),
            response_model=QueryResponse,
        )
    
    async def validate(self, query: str) -> Dict[str, Any]:
        """Validate a query without executing it."""
        data = await self.client.post("/query/validate", json={"query": query})
        return data
    
    async def suggest(
        self,
        partial: str,
        event_type: Optional[str] = None,
        cursor_position: Optional[int] = None,
    ) -> Dict[str, List[Dict[str, Any]]]:
        """Get query suggestions based on partial input."""
        request_data = {"partial": partial}
        if event_type:
            request_data["eventType"] = event_type
        if cursor_position is not None:
            request_data["cursorPosition"] = cursor_position
        
        data = await self.client.post("/query/suggest", json=request_data)
        return data
    
    async def format(self, query: str) -> Dict[str, Any]:
        """Format a query."""
        data = await self.client.post("/query/format", json={"query": query})
        return data
    
    async def get_history(
        self,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        from_time: Optional[str] = None,
        to_time: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get query history."""
        params = {}
        if limit is not None:
            params["limit"] = str(limit)
        if offset is not None:
            params["offset"] = str(offset)
        if from_time:
            params["from"] = from_time
        if to_time:
            params["to"] = to_time
        
        data = await self.client.get("/query/history", params=params)
        return data
    
    async def export(
        self, query_id: str, format: str = "csv"
    ) -> bytes:
        """Export query results."""
        response = await self.client.request(
            "GET",
            f"/query/{quote(query_id, safe='')}/export",
            params={"format": format},
        )
        return response.content


class SyncQueryService:
    """Synchronous service for query operations."""
    
    def __init__(self, client: "SyncUDSClient"):
        self.client = client
    
    def execute(self, request: QueryRequest) -> QueryResponse:
        """Execute a query."""
        return self.client.post(
            "/query",
            json=request.model_dump(by_alias=True, exclude_none=True),
            response_model=QueryResponse,
        )
    
    def validate(self, query: str) -> Dict[str, Any]:
        """Validate a query without executing it."""
        data = self.client.post("/query/validate", json={"query": query})
        return data
    
    def suggest(
        self,
        partial: str,
        event_type: Optional[str] = None,
        cursor_position: Optional[int] = None,
    ) -> Dict[str, List[Dict[str, Any]]]:
        """Get query suggestions based on partial input."""
        request_data = {"partial": partial}
        if event_type:
            request_data["eventType"] = event_type
        if cursor_position is not None:
            request_data["cursorPosition"] = cursor_position
        
        data = self.client.post("/query/suggest", json=request_data)
        return data
    
    def format(self, query: str) -> Dict[str, Any]:
        """Format a query."""
        data = self.client.post("/query/format", json={"query": query})
        return data
    
    def get_history(
        self,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        from_time: Optional[str] = None,
        to_time: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get query history."""
        params = {}
        if limit is not None:
            params["limit"] = str(limit)
        if offset is not None:
            params["offset"] = str(offset)
        if from_time:
            params["from"] = from_time
        if to_time:
            params["to"] = to_time
        
        data = self.client.get("/query/history", params=params)
        return data
    
    def export(self, query_id: str, format: str = "csv") -> bytes:
        """Export query results."""
        response = self.client.request(
            "GET",
            f"/query/{quote(query_id, safe='')}/export",
            params={"format": format},
        )
        return response.content